File: LeetCode/Medium/2130-maximum-twin-sum-of-a-linked-list/2130-maximum-twin-sum-of-a-linked-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public int pairSum(ListNode head) {
        
        int idx = 1;
        ListNode temp = head;
        while(temp.next.next != null) {
            temp = temp.next.next;
            idx++;
        }

        int ans = 0;
        int[] stack = new int[idx + 1];
        temp = head;
        int s = 1;
        boolean check = true;
        while(temp != null) {
            if(check) {
                stack[s] = temp.val;
                if(s != idx) s++;
                else {
                    check = false;

                }
            } else {
                ans = Math.max(ans, stack[s--] + temp.val);
            }
            

            temp = temp.next;
        }
        return ans;
    }
}

File: LeetCode/Medium/2130-maximum-twin-sum-of-a-linked-list/2130-maximum-twin-sum-of-a-linked-list.kt
/**
 * Example:
 * var li = ListNode(5)
 * var v = li.`val`
 * Definition for singly-linked list.
 * class ListNode(var `val`: Int) {
 *     var next: ListNode? = null
 * }
 */
class Solution {
    fun pairSum(head: ListNode?): Int {

        if(head == null) return 0

        var ans = 0

        var fast = head
        var slow = head
        var curr = head
        var prev: ListNode? = null

        while(fast != null) {
            fast = fast.next.next
            slow = slow!!.next
            curr!!.next = prev
            prev = curr
            curr = slow
            
        }

        while(slow != null) {
            ans = Math.max(ans, prev!!.`val` + slow!!.`val`)
            prev = prev.next
            slow = slow.next
        }

        return ans
        
    }
}